crates/wasmtime/src/runtime/vm/component.rs

```diff
@@ -825,22 +851,30 @@ impl ComponentInstance {
         resource_instance == component.defined_resource_instances[idx]
     }
 
-    /// Returns the runtime state of resources associated with this component.
+    /// Returns the runtime state of resources and concurrency associated with
+    /// this component.
     #[inline]
-    pub fn guest_tables(
+    pub fn instance_states(
         self: Pin<&mut Self>,
     ) -> (
-        &mut PrimaryMap<RuntimeComponentInstanceIndex, HandleTable>,
+        &mut PrimaryMap<RuntimeComponentInstanceIndex, InstanceState>,
         &ComponentTypes,
     ) {
         // safety: we've chosen the `pin` guarantee of `self` to not apply to
         // the map returned.
         unsafe {
             let me = self.get_unchecked_mut();
-            (&mut me.instance_handle_tables, me.component.types())
+            (&mut me.instance_states, me.component.types())
         }
     }
 
+    pub fn instance_state(
+        self: Pin<&mut Self>,
+        instance: RuntimeComponentInstanceIndex,
+    ) -> Option<&mut InstanceState> {
+        self.instance_states().0.get_mut(instance)
+    }
```

Comment:
For this almost all other `*Index`-based accessors panic internally on out-of-bounds, so I think it's reasonable to return `&mut` here instead of `Option<&mut T>

Reply:
Ugh, I forgot to address this before merging.  I'll address it in a follow-up.